### PR TITLE
per-message compression options

### LINKED ISSUE #226: allow per-message compression option
When permessage-deflate is negotiated on a stream, It should be possible to control whether compression is enabled on a per-message basis.

### PR DESCRIPTION
**Pull request by alandefreitas**
fix #226, #227

### CODE REVIEW

File: include/boost/beast/websocket/option.hpp [hunk 1]
```diff
@@ -55,6 +55,9 @@ struct permessage_deflate
 
     /// Deflate memory level, 1..9
     int memLevel = 4;
+
+    /// The minimum size a message should have to be compressed
+    std::size_t msg_size_threshold = 0;
```
Comment (alandefreitas) [inline include/boost/beast/websocket/option.hpp:60]: I couldn't find a reference for a reasonable default yet...

File: include/boost/beast/websocket/impl/write.hpp [hunk 2]
```diff
@@ -86,7 +86,7 @@ class stream<NextLayer, deflateSupported>::write_some_op
         // Set up the outgoing frame header
         if(! impl.wr_cont)
         {
-            impl.begin_msg();
+            impl.begin_msg(buffer_bytes(bs));
```
Comment (vinniefalco) [inline include/boost/beast/websocket/impl/write.hpp:89]: You need to inhibit ADL here, because `bs` is user-defined

File: include/boost/beast/websocket/impl/write.hpp [hunk 3]
```diff
@@ -303,7 +303,7 @@ operator()(
         if(how_ == do_mask_nofrag)
         {
             // send a single frame using multiple writes
-            remain_ = beast::buffer_bytes(cb_);
+            remain_ = buffer_bytes(cb_);
```
Comment (vinniefalco) [inline include/boost/beast/websocket/impl/write.hpp:306]: You need to inhibit ADL here

File: include/boost/beast/websocket/impl/write.hpp [hunk 4]
```diff
@@ -622,7 +622,7 @@ write_some(bool fin,
     detail::frame_header fh;
     if(! impl.wr_cont)
     {
-        impl.begin_msg();
+        impl.begin_msg(buffer_bytes(buffers));
```
Comment (vinniefalco) [inline include/boost/beast/websocket/impl/write.hpp:625]: you might need to inhibit ADL here

File: include/boost/beast/websocket/stream_fwd.hpp [hunk 5]
```diff
@@ -12,18 +12,18 @@
...
 
+#ifndef BOOST_BEAST_DOXYGEN
+
 //[code_websocket_1h
 
 namespace boost {
 namespace beast {
 namespace websocket {
 
-#ifndef BOOST_BEAST_DOXYGEN
```
Comment (vinniefalco) [inline include/boost/beast/websocket/stream_fwd.hpp:21]: Why did you remove the ifdef? This will probably confuse doxygen

Comment (alandefreitas) [reply include/boost/beast/websocket/stream_fwd.hpp:21]: I didn't remove it. I just moved it outside the quickbook comment to fix this small bug in the docs that were rendering `BOOST_BEAST_DOXYGEN` sometimes.

image

To be honest, I just checked https://www.boost.org/doc/libs/1_79_0/libs/beast/doc/html/beast/using_websocket.html and it seems like this only happens locally. I have no idea why.

**Comment by klemens-morgenstern** [comment]
Looks to be failing due to #2481.

**Comment by klemens-morgenstern** [comment]
Drone passes now, the 4.8 issues seems fixed. @vinniefalco where are we at with this?

**Comment by vinniefalco** [comment]
Should be ok

**Comment by klemens-morgenstern** [comment]
Commited directly as squash
